### PR TITLE
layout: add Access Plan API

### PR DESCRIPTION
**Pull request by andriytk**
Designed by Nikita in 2016, but never implemented. Maybe it's a good time to revive the idea as it might be useful for the In-Storage Compute (ISC, aka Function Shipping).

### CODE REVIEW

File: layout/plan.h
Comment (andriytk) [inline layout/plan.h:51]: @nikitadanilov what are these structs and functions in the beginning and how they are related to the rest of the API?

Comment (siningwuseagate) [reply layout/plan.h:51]: @andriytk  here is something I remember when implementing composite layout (for Sage1 project): a composite object operation usually needs to issue a few ops,  such as querying extents from index and then sending IO ops, currently the composite object code does these ops in sequential steps, meaning we have to wait for (or block on) one step until we can move to next steps. Another is for the error handling in composite layout, when one op fails, how to deal with the whole IO.  Back then when implementing composite layout, layout plan was one of the option, but implementing it means the whole clovis IO code had to be re-written. So Sage1 didn't do it.

If we're going to implement the layout plan and use it for isc only, then motr may have 2 set of code doing similar thing (motr client IO code and layout plan), I am not sure if it is a good idea. But if motr is going to refactor client side code to use layout plan, that is going to be a huge job.

**Comment by johnbent** [comment]
Plus @TechWriter-Mayur who I believe has already added this same plan.h somewhere else in cortx-motr repo.  Mayur, I trust Andriy to know the best logical place to add this file.  Can you please watch this PR and after it gets merged, can you remove your plan.h and change the links that you made to point to this new one?

File: layout/plan.h
Comment (huanghua78) [inline layout/plan.h:31]: m0_layout is already defined in "layout/layout.h", and please note, the two definitions are different.

Can we just have a forward declaration here?

Comment (huanghua78) [reply layout/plan.h:31]: I know this file has not been included in the Makefile. But removing these obvious bug seems necessary.

**Review by huanghua78 — APPROVED** [review]

File: layout/plan.h
Comment (andriytk) [reply layout/plan.h:31]: @huanghua78, I asked @nikitadanilov above about these structs, let's hear from him. Of course, we shouldn't land it as it is.

File: layout/plan.h
Comment (andriytk) [inline layout/plan.h:145]: As appeared, there will be always `M0_LAT_OUT_READ` after `M0_LAT_READ` to make the data available for user.

File: layout/plan.h
Comment (andriytk) [reply layout/plan.h:51]: According to @nikitadanilov these two structures and functions were added by mistake to this file, and they just can be removed.

File: layout/plan.h
Comment (andriytk) [inline layout/plan.h:111]: @nikitadanilov how Function Shipping code (user) can know whether we are going to read the real data or the parity data units? FS needs to know where the data units are located in order to send the computation requests there.

Comment (andriytk) [reply layout/plan.h:111]: @nikitadanilov, what if we introduce an additional plop type - `M0_LAT_MUST_READ` which would indicate that the user must try to read the actual data from the server (data or parity unit) and put it in `iop_data` when calling plop_done()? The simple `M0_LAT_READ` would indicate that the plan is working in normal mode (not the degraded one), so the ISC code could just send the computing request to the server without reading the actual data.

Looks like it would solve the problems for ISC mentioned above.

Comment (nikitadanilov) [reply layout/plan.h:111]: This only works for a particular combination of layout type (parity de-clustered layout type) and particular user (ISC). I don't see how he distinction between `READ` and `MUST_READ` makes sense for an arbitrary user of an arbitrary layout type.

I think the notion of layout plan as a graph (represented by dependencies) is generic enough to capture ISC and other users including normal IO. Let's discuss this on a call. I'll leave two dot-graphs for a normal read and a read with failure as examples.

```digraph plop {
	subgraph cluster_0 {
		style=filled;
		color=lightgrey;
		start;
		getlayout [label="layout-index.GET(fid)"];
		getlayout -> start;
	}
	subgraph cluster_1 {
		style=filled;
		color=lightgrey;
		read0 [label="cob0.read(ext0)"];
		read1 [label="cob1.read(ext1)"];
		read2 [label="cob2.read(ext2)"];
		readout0 [label="buf0.ready"];
		readout1 [label="buf1.ready"];
		readout2 [label="buf2.ready"];
		readout0 -> read0;
		readout1 -> read1;
		readout2 -> read2;
		read0 -> getlayout [style=invis];
		read1 -> getlayout [style=invis];
		read2 -> getlayout [style=invis];
	}
	subgraph cluster_2 {
		style=filled;
		color=lightgrey;
		done
		done -> readout0 [style=invis];
		done -> readout1 [style=invis];
		done -> readout2 [style=invis];
	}
}
```

```
digraph plopfail {
	subgraph cluster_0 {
		style=filled;
		color=lightgrey;
		start;
		getlayout [label="layout-index.GET(fid)"];
		getlayout -> start;
	}
	subgraph cluster_1 {
		style=filled;
		color=lightgrey;
		read0 [label="cob0.read(ext0)" color="red"];
		read1 [label="cob1.read(ext1)"];
		read2 [label="cob2.read(ext2)"];
		readout0 [label="buf0.ready" color="blue"];
		readout1 [label="buf1.ready"];
		readout2 [label="buf2.ready"];
		readout0 -> read0;
		readout1 -> read1;
		readout2 -> read2;
		read0 -> getlayout [style=invis];
		read1 -> getlayout [style=invis];
		read2 -> getlayout [style=invis];
	}
	subgraph cluster_degraded {
		style=filled;
		color=lightgrey;
		readparity [label="cob-parity.read(ext-parity)"];
		readparity -> read0 [style=invis];
		compute [label="erasure(buf1, buf2, buf-parity)"];
		readout0p [label="buf0.ready"];
		compute -> readparity;
		compute -> read1;
		compute -> read2;
		readout0p -> compute;
	}
	subgraph cluster_2 {
		style=filled;
		color=lightgrey;
		done
		done -> readout0p [style=invis];
		done -> readout1 [style=invis];
		done -> readout2 [style=invis];
	}
}
```

Comment (andriytk) [reply layout/plan.h:111]: > I don't see how he distinction between READ and MUST_READ makes sense for an arbitrary user of an arbitrary layout type.

@nikitadanilov, could you please elaborate this statement a bit? For this distinction looks generic enough for ISC needs when a clear indication from the implementation is needed whether the request for the computation can be sent to the server side to be done closer to where the data is stored or not.

Comment (andriytk) [reply layout/plan.h:111]: After discussion with @nikitadanilov there seem to be two major issues with the proposed approach: 1) it won't allow using additional layouts for ISC (like encryption or checksums); 2) it introduces double-reading of the parity group units in the normal i/o path in case of an incidental unforeseeable degraded read (when some READ returns error). Therefore, here is an improved proposition which tackles the above issues and still allows to keep the interface simple for the users:

1. Return READ (instead of MUST_READ) even if there are some FUNctions that must be executed on the read unit data (like checksumming or decompression), but all these functions don't require any other READs in order to get the resulting data for OUT_READ. (Straight line dependency in the graph.) This would tell ISC that the processing of this unit can be done at the server side.

2. Introduce m0_layout_plop_ops::po_update(plop) callback which is called when READ is converted to MUST_READ. This would allow the i/o user to avoid the double-reads on an incidental fallback to degraded read of a parity group. (Note: it is assumed that the OUT_READ plop is returned with parity group granularity and the READ plop is not destroyed until its resulting OUT_READ is ready. This guarantees that po_update(plop) can always be called when needed.)

Comment (andriytk) [reply layout/plan.h:111]: See the proposition in detail at commit https://github.com/Seagate/cortx-motr/pull/411/commits/9fc08d67b01e12f01d6ceb937e3a9753b3286626.

Comment (nikitadanilov) [reply layout/plan.h:111]: As I mentioned during the discussion, it is not clear how the proposal handles the difference between compression (that can be handled by ISC) and client-only encryption (that cannot be handled by ISC).

The only way to define `MUST_READ` precisely and generically, is to define it as `CANNOT_DELEGATE`.  I object to the idea that ISC-specific analysis (that results in the distinction between `READ` and `MUST_READ` plops) should be done by the plan implementation (that is, should be implemented separately by each layout type). This ISC-specific analysis should be done by ISC.

Also, not enough details are provided for the proposed ->po_update() call-back:

    - how the user is supposed to know which plop attributes changed?

    - "The implementation invokes this for a still not complete plop..."---what if the user already completed the `READ` plop during normal IO and already freed the buffer? More generally, what are the ownership and liveness assumptions for plops and buffers in this proposal?

File: layout/plan.h
Comment (huanghua78) [inline layout/plan.h:58]: Do you mean, when the "Access plan" APIs are ready, we need some kind of re-write of client code?

File: layout/plan.h
Comment (huanghua78) [inline layout/plan.h:305]: s/M0_LAT_IN_WRITE/M0_LPT_IN_WRITE
?

similar to M0_LPS_INIT.

File: layout/plan.h
Comment (huanghua78) [inline layout/plan.h:412]: Please clarify the link list head which the link is linked to.

File: layout/plan.h
Comment (huanghua78) [reply layout/plan.h:58]: As told in the introduction PDF, client i/o code may be re-written to use these APIs.

Comment (andriytk) [reply layout/plan.h:58]: Yes, it can be rewritten later when the plan API implementation is ready and stable.

File: layout/plan.h
Comment (andriytk) [reply layout/plan.h:305]: Fine with me, but it's the original naming from @nikitadanilov. We can update if it's fine with him.

File: layout/plan.h
Comment (andriytk) [reply layout/plan.h:412]: It's probably in the `m0_layout_plan` structure which will be introduced later during the implementation. Probably in a `plan_internal.h` file. I will update it then.

**Comment by huanghua78** [comment]
@nikitadanilov , can you please review this PR?

**Comment by huanghua78** [comment]
@nikitadanilov , can you please review this PR?

**Review by nikitadanilov — APPROVED** [review]

**Pull request merged**
